src/lib/communication/DY.Lib.Communication.Core.fst

```diff
@@ -60,110 +65,105 @@ instance event_communication_event: event communication_event = {
 
 [@@ "opaque_to_smt"]
 val encrypt_message:
-  #a:Type -> {| parseable_serializeable bytes a |} ->
+  #a:Type -> {|comm_layer_core_config a|} ->
   bytes -> bytes -> a -> bytes
 let encrypt_message #a pk_receiver nonce payload =
   pke_enc pk_receiver nonce (serialize a payload)
 
 [@@ "opaque_to_smt"]
 val send_confidential:
-  #a:Type -> {| parseable_serializeable bytes a |} ->
+  #a:Type0 -> {|comm_layer_core_config a|} ->
   communication_keys_sess_ids ->
   principal -> principal -> a ->
   traceful (option timestamp)
 let send_confidential #a comm_keys_ids sender receiver payload =
-  let*? pk_receiver = get_public_key sender comm_keys_ids.pki (LongTermPkeKey comm_layer_pkenc_tag) receiver in
+  let*? pk_receiver = get_public_key sender comm_keys_ids.pki (LongTermPkeKey (comm_layer_pkenc_tag a)) receiver in
   let* nonce = mk_rand PkeNonce (long_term_key_label sender) 32 in
-  trigger_event sender (CommConfSendMsg sender receiver (serialize a payload));*
+  trigger_event sender (CommConfSendMsg sender receiver payload <: communication_core_event a);*
   let msg_encrypted = encrypt_message pk_receiver nonce payload in
   let* msg_id = send_msg msg_encrypted in
   return (Some msg_id)
 
 [@@ "opaque_to_smt"]
 val decrypt_message:
-  #a:Type -> {| parseable_serializeable bytes a |} ->
+  #a:Type0 -> {|comm_layer_core_config a|}  ->
   bytes -> bytes -> option a
 let decrypt_message #a sk_receiver msg_encrypted =
   let? plaintext = pke_dec sk_receiver msg_encrypted in
   parse #bytes a plaintext
 
 [@@ "opaque_to_smt"]
 val receive_confidential:
-  #a:Type -> {| parseable_serializeable bytes a |} ->
+  #a:Type0 -> {|comm_layer_core_config a|}  ->
   communication_keys_sess_ids ->
   principal -> timestamp ->
   traceful (option a)
 let receive_confidential #a comm_keys_ids receiver msg_id =
-  let*? sk_receiver = get_private_key receiver comm_keys_ids.private_keys (LongTermPkeKey comm_layer_pkenc_tag) in
+  let*? sk_receiver = get_private_key receiver comm_keys_ids.private_keys (LongTermPkeKey (comm_layer_pkenc_tag a)) in
   let*? msg_encrypted = recv_msg msg_id in
-  let*? payload = return (decrypt_message #a sk_receiver msg_encrypted) in
-  trigger_event receiver (CommConfReceiveMsg receiver (serialize a payload));*
+  let*? payload = return (decrypt_message sk_receiver msg_encrypted) in
+  trigger_event receiver (CommConfReceiveMsg receiver payload <: communication_core_event a);*
   return (Some payload)
 
 
 (**** Authenticated Send and Receive Functions ****)
 
-#push-options "--ifuel 1"
 [@@ "opaque_to_smt"]
 val sign_message:
-  #a:Type -> {| parseable_serializeable bytes a |} ->
-  principal -> principal -> a -> option bytes -> bytes -> bytes -> bytes
+  #a:Type -> {|comm_layer_core_config a|} ->
+  principal -> principal -> payload:either a bytes -> pk_receiver:option bytes{(Some? pk_receiver <==> Inr? payload) /\ (None? pk_receiver <==> Inl? payload)} -> bytes -> bytes -> bytes
```

Comment:
I wonder if it makes sense to bundle the payload and `pk_receiver` into a single type? Something of the form `either a (bytes & bytes)`, though maybe with a nicer name given to the right case and some form of record to extract things properly --- `either a ({payload:bytes; pk_receiver:bytes})` or so?